File: java/Simulation/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import static edu.wpi.first.units.Units.*;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusCode;
import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.Pigeon2Configuration;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.hardware.CANcoder;
import com.ctre.phoenix6.hardware.Pigeon2;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.sim.CANcoderSimState;
import com.ctre.phoenix6.sim.ChassisReference;
import com.ctre.phoenix6.sim.Pigeon2SimState;
import com.ctre.phoenix6.sim.TalonFXSimState;
import com.ctre.phoenix6.signals.SensorDirectionValue;
import com.ctre.phoenix6.signals.InvertedValue;

import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.units.measure.LinearVelocity;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.simulation.DifferentialDrivetrainSim;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to
 * each mode, as described in the TimedRobot documentation. If you change the
 * name of this class or
 * the package after creating this project, you must also update the
 * build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot {
    private final XboxController joystick = new XboxController(0);

    private final TalonFX leftFX = new TalonFX(1);
    private final TalonFX rightFX = new TalonFX(2);
    private final CANcoder leftSensor = new CANcoder(1);
    private final CANcoder rightSensor = new CANcoder(2);
    private final Pigeon2 imu = new Pigeon2(0);

    private final TalonFXSimState leftSim = leftFX.getSimState();
    private final TalonFXSimState rightSim = rightFX.getSimState();
    private final CANcoderSimState leftSensSim = leftSensor.getSimState();
    private final CANcoderSimState rightSensSim = rightSensor.getSimState();
    private final Pigeon2SimState imuSim = imu.getSimState();

    private final DifferentialDrive drivetrain = new DifferentialDrive(leftFX, rightFX);

    /*
     * These numbers are an example AndyMark Drivetrain with some additional weight.
     * This is a fairly light robot.
     * Note you can utilize results from robot characterization instead of
     * theoretical numbers.
     * https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-
     * characterization/introduction.html#introduction-to-robot-characterization
     */
    private final double kGearRatio = 10.71;
    private final Distance kWheelRadius = Inches.of(3);

    /* Simulation model of the drivetrain */
    private final DifferentialDrivetrainSim m_driveSim = new DifferentialDrivetrainSim(
        DCMotor.getFalcon500Foc(2), // 2 CIMS on each side of the drivetrain.
        kGearRatio, // Standard AndyMark Gearing reduction.
        2.1, // MOI of 2.1 kg m^2 (from CAD model).
        26.5, // Mass of the robot is 26.5 kg.
        kWheelRadius.in(Inches), // Robot uses 3" radius (6" diameter) wheels.
        0.546, // Distance between wheels is _ meters.

        /*
         * The standard deviations for measurement noise:
         * x and y: 0.001 m
         * heading: 0.001 rad
         * l and r velocity: 0.1 m/s
         * l and r position: 0.005 m
         */
        /* Uncomment the following line to add measurement noise. */
        null // VecBuilder.fill(0.001, 0.001, 0.001, 0.1, 0.1, 0.005, 0.005)
    );

    private final Field2d m_field = new Field2d();
    /*
     * Creating my odometry object.
     */
    private final DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(
        imu.getRotation2d(),
        0, 0
    );

    /**
     * This function is run when the robot is first started up and should be used
     * for any
     * initialization code.
     */
    @Override
    public void robotInit() {
        StatusCode returnCode;

        TalonFXConfiguration fxCfg = new TalonFXConfiguration();
        fxCfg.MotorOutput.Inverted = InvertedValue.CounterClockwise_Positive;
        for (int i = 0; i < 5; ++i) {
            returnCode = leftFX.getConfigurator().apply(fxCfg);
            if (returnCode.isOK()) break;
        }

        fxCfg.MotorOutput.Inverted = InvertedValue.Clockwise_Positive;
        for (int i = 0; i < 5; ++i) {
            returnCode = rightFX.getConfigurator().apply(fxCfg);
            if (returnCode.isOK()) break;
        };

        CANcoderConfiguration ccCfg = new CANcoderConfiguration();
        ccCfg.MagnetSensor.SensorDirection = SensorDirectionValue.CounterClockwise_Positive;
        leftSensor.getConfigurator().apply(ccCfg);
        ccCfg.MagnetSensor.SensorDirection = SensorDirectionValue.Clockwise_Positive;
        rightSensor.getConfigurator().apply(ccCfg);

        Pigeon2Configuration imuCfg = new Pigeon2Configuration();
        imu.getConfigurator().apply(imuCfg);

        /* Make sure all critical signals are synchronized */
        /*
         * Setting all these signals to 100hz means they get sent at the same time if
         * they're all on a CANivore
         */
        BaseStatusSignal.setUpdateFrequencyForAll(100,
            leftFX.getPosition(),
            rightFX.getPosition(),
            imu.getYaw());

        /* Publish field pose data to read back from */
        SmartDashboard.putData("Field", m_field);
    }

    private int printCount = 0;

    @Override
    public void robotPeriodic() {
        /*
         * This will get the simulated sensor readings that we set
         * in the previous article while in simulation, but will use
         * real values on the robot itself.
         */
        m_odometry.update(imu.getRotation2d(),
                rotationsToMeters(leftSensor.getPosition().getValue()).in(Meters),
                rotationsToMeters(rightSensor.getPosition().getValue()).in(Meters));
        m_field.setRobotPose(m_odometry.getPoseMeters());

        if (++printCount >= 50) {
            printCount = 0;
            System.out.println("Left FX: " + leftFX.getPosition());
            System.out.println("Right FX: " + rightFX.getPosition());
            System.out.println("Left CANcoder: " + leftSensor.getPosition());
            System.out.println("Right CANcoder: " + rightSensor.getPosition());
            System.out.println("Left Forward limit: " + leftFX.getForwardLimit());
            System.out.println("Left Reverse limit: " + leftFX.getReverseLimit());
            System.out.println("Right Forward limit: " + rightFX.getForwardLimit());
            System.out.println("Right Reverse limit: " + rightFX.getReverseLimit());
            System.out.println("Pigeon2: " + imu.getYaw());
            System.out.println();
        }
    }

    @Override
    public void simulationInit() {
        /*
         * Set the orientation of the simulated devices relative to the robot chassis.
         * WPILib expects +V to be forward. Specify orientations to match that behavior.
         */
        /* left devices are CCW+ */
        leftSim.Orientation = ChassisReference.CounterClockwise_Positive;
        leftSensSim.Orientation = ChassisReference.CounterClockwise_Positive;
        /* right devices are CW+ */
        rightSim.Orientation = ChassisReference.Clockwise_Positive;
        rightSensSim.Orientation = ChassisReference.Clockwise_Positive;
    }

    @Override
    public void simulationPeriodic() {
        /* Pass the robot battery voltage to the simulated devices */
        leftSim.setSupplyVoltage(RobotController.getBatteryVoltage());
        leftSensSim.setSupplyVoltage(RobotController.getBatteryVoltage());
        rightSim.setSupplyVoltage(RobotController.getBatteryVoltage());
        rightSensSim.setSupplyVoltage(RobotController.getBatteryVoltage());
        imuSim.setSupplyVoltage(RobotController.getBatteryVoltage());

        /*
         * CTRE simulation is low-level, so SimState inputs
         * and outputs are not affected by user-level inversion.
         * However, inputs and outputs *are* affected by the mechanical
         * orientation of the device relative to the robot chassis,
         * as specified by the `orientation` field.
         *
         * WPILib expects +V to be forward. We have already configured
         * our orientations to match this behavior.
         */
        m_driveSim.setInputs(leftSim.getMotorVoltage(),
                rightSim.getMotorVoltage());

        /*
         * Advance the model by 20 ms. Note that if you are running this
         * subsystem in a separate thread or have changed the nominal
         * timestep of TimedRobot, this value needs to match it.
         */
        m_driveSim.update(0.02);

        /* Update all of our sensors. */
        final var leftPos = metersToRotations(
                Meters.of(m_driveSim.getLeftPositionMeters()));
        // This is OK, since the time base is the same
        final var leftVel = metersToRotationsVel(
                MetersPerSecond.of(m_driveSim.getLeftVelocityMetersPerSecond()));
        final var rightPos = metersToRotations(
                Meters.of(m_driveSim.getRightPositionMeters()));
        // This is OK, since the time base is the same
        final var rightVel = metersToRotationsVel(
                MetersPerSecond.of(m_driveSim.getRightVelocityMetersPerSecond()));
        leftSensSim.setRawPosition(leftPos);
        leftSensSim.setVelocity(leftVel);
        rightSensSim.setRawPosition(rightPos);
        rightSensSim.setVelocity(rightVel);
        leftSim.setRawRotorPosition(leftPos.times(this.kGearRatio));
        leftSim.setRotorVelocity(leftVel.times(this.kGearRatio));
        rightSim.setRawRotorPosition(rightPos.times(this.kGearRatio));
        rightSim.setRotorVelocity(rightVel.times(this.kGearRatio));
        imuSim.setRawYaw(m_driveSim.getHeading().getDegrees());


        /*
         * If a bumper is pressed, trigger the forward limit switch to test it, 
         * if a trigger is pressed, trigger the reverse limit switch 
         */
        leftSim.setForwardLimit(joystick.getLeftBumperButton());
        leftSim.setReverseLimit(joystick.getLeftTriggerAxis() > 0.5);
        rightSim.setForwardLimit(joystick.getRightBumperButton());
        rightSim.setReverseLimit(joystick.getRightTriggerAxis() > 0.5);
    }

    @Override
    public void autonomousInit() {
    }

    @Override
    public void autonomousPeriodic() {
    }

    @Override
    public void teleopInit() {
    }

    @Override
    public void teleopPeriodic() {
        drivetrain.curvatureDrive(-joystick.getLeftY(), -joystick.getRightX(), joystick.getRightStickButton());
    }

    @Override
    public void disabledInit() {
    }

    @Override
    public void disabledPeriodic() {
    }

    @Override
    public void testInit() {
    }

    @Override
    public void testPeriodic() {
    }

    private Distance rotationsToMeters(Angle rotations) {
        /* Apply gear ratio to input rotations */
        var gearedRadians = rotations.in(Radians) / this.kGearRatio;
        /* Then multiply the wheel radius by radians of rotation to get distance */
        return this.kWheelRadius.times(gearedRadians);
    }

    private Angle metersToRotations(Distance meters) {
        /* Divide the distance by the wheel radius to get radians */
        var wheelRadians = meters.in(Meters) / this.kWheelRadius.in(Meters);
        /* Then multiply by gear ratio to get rotor rotations */
        return Radians.of(wheelRadians * this.kGearRatio);
    }

    private LinearVelocity rotationsToMetersVel(AngularVelocity rotations) {
        /* Apply gear ratio to input rotations */
        var gearedRotations = rotations.divide(this.kGearRatio);
        /* Then multiply the wheel radius by radians of rotation to get distance */
        return this.kWheelRadius.per(Second).times(gearedRotations.in(RadiansPerSecond));
    }

    private AngularVelocity metersToRotationsVel(LinearVelocity meters) {
        /* Divide the distance by the wheel radius to get radians */
        var wheelRadians = meters.in(MetersPerSecond) / this.kWheelRadius.in(Meters);
        /* Then multiply by gear ratio to get rotor rotations */
        return RadiansPerSecond.of(wheelRadians * this.kGearRatio);
    }
}

File: cpp/Simulation/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/RobotController.h>
#include <iostream>

using namespace ctre::phoenix6;

void Robot::RobotInit()
{
  ctre::phoenix::StatusCode returnCode;

  configs::TalonFXConfiguration fxCfg{};
  fxCfg.MotorOutput.Inverted = signals::InvertedValue::CounterClockwise_Positive;
  int retryCount = 5;
  for (int i = 0; i < retryCount; ++i) {
    returnCode = leftFX.GetConfigurator().Apply(fxCfg);
    if (returnCode.IsOK()) break;
  }

  fxCfg.MotorOutput.Inverted = signals::InvertedValue::Clockwise_Positive;
  retryCount = 5;
  for (int i = 0; i < retryCount; ++i) {
    returnCode = rightFX.GetConfigurator().Apply(fxCfg);
    if (returnCode.IsOK()) break;
  }

  configs::CANcoderConfiguration ccCfg{};
  ccCfg.MagnetSensor.SensorDirection = signals::SensorDirectionValue::CounterClockwise_Positive;
  leftSensor.GetConfigurator().Apply(ccCfg);
  ccCfg.MagnetSensor.SensorDirection = signals::SensorDirectionValue::Clockwise_Positive;
  rightSensor.GetConfigurator().Apply(ccCfg);

  configs::Pigeon2Configuration imuCfg{};
  imu.GetConfigurator().Apply(imuCfg);

  /* Make sure all critical signals are synchronized */
  /*
   * Setting all these signals to 100hz means they get sent at the same time if
   * they're all on a CANivore
   */
  BaseStatusSignal::SetUpdateFrequencyForAll(100_Hz,
      leftFX.GetPosition(),
      rightFX.GetPosition(),
      imu.GetYaw());

  /* Publish field pose data to read back from */
  frc::SmartDashboard::PutData("Field", &m_field);
}

void Robot::RobotPeriodic()
{
  /*
   * This will get the simulated sensor readings that we set
   * in the previous article while in simulation, but will use
   * real values on the robot itself.
   */
  m_odometry.Update(imu.GetRotation2d(),
                    rotationsToMeters(leftSensor.GetPosition().GetValue()),
                    rotationsToMeters(rightSensor.GetPosition().GetValue()));
  m_field.SetRobotPose(m_odometry.GetPose());

  if (++printCount >= 50) {
    printCount = 0;
    std::cout << "Left FX: " << leftFX.GetPosition() << std::endl;
    std::cout << "Right FX: " << rightFX.GetPosition() << std::endl;
    std::cout << "Left CANcoder: " << leftSensor.GetPosition() << std::endl;
    std::cout << "Right CANcoder: " << rightSensor.GetPosition() << std::endl;
    std::cout << "Left Forward limit: " << leftFX.GetForwardLimit() << std::endl;
    std::cout << "Left Reverse limit: " << leftFX.GetReverseLimit() << std::endl;
    std::cout << "Right Forward limit: " << rightFX.GetForwardLimit() << std::endl;
    std::cout << "Right Reverse limit: " << rightFX.GetReverseLimit() << std::endl;
    std::cout << "Pigeon2: " << imu.GetYaw() << std::endl;
    std::cout << "" << std::endl;
  }
}

void Robot::AutonomousInit() {}
void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {}
void Robot::TeleopPeriodic()
{
  drivetrain.CurvatureDrive(-joystick.GetLeftY(), -joystick.GetRightX(), joystick.GetRightStickButton());
}

void Robot::DisabledInit() {}
void Robot::DisabledPeriodic() {}

void Robot::TestInit() {}
void Robot::TestPeriodic() {}

void Robot::SimulationInit()
{
  /*
   * Set the orientation of the simulated devices relative to the robot chassis.
   * WPILib expects +V to be forward. Specify orientations to match that behavior.
   */
  /* left devices are CCW+ */
  leftSim.Orientation = sim::ChassisReference::CounterClockwise_Positive;
  leftSensSim.Orientation = sim::ChassisReference::CounterClockwise_Positive;
  /* right devices are CW+ */
  rightSim.Orientation = sim::ChassisReference::Clockwise_Positive;
  rightSensSim.Orientation = sim::ChassisReference::Clockwise_Positive;
}
void Robot::SimulationPeriodic()
{
  /* Pass the robot battery voltage to the simulated devices */
  leftSim.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
  leftSensSim.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
  rightSim.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
  rightSensSim.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
  imuSim.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());

  /*
   * CTRE simulation is low-level, so SimState inputs
   * and outputs are not affected by user-level inversion.
   * However, inputs and outputs *are* affected by the mechanical
   * orientation of the device relative to the robot chassis,
   * as specified by the `orientation` field.
   *
   * WPILib expects +V to be forward. We have already configured
   * our orientations to match this behavior.
   */
  m_driveSim.SetInputs(leftSim.GetMotorVoltage(),
                       rightSim.GetMotorVoltage());

  /*
   * Advance the model by 20 ms. Note that if you are running this
   * subsystem in a separate thread or have changed the nominal
   * timestep of TimedRobot, this value needs to match it.
   */
  m_driveSim.Update(20_ms);

  /* Update all of our sensors. */
  auto const leftPos = metersToRotations(m_driveSim.GetLeftPosition());
  auto const leftVel = metersToRotationsVel(m_driveSim.GetLeftVelocity());
  auto const rightPos = metersToRotations(m_driveSim.GetRightPosition());
  auto const rightVel = metersToRotationsVel(m_driveSim.GetRightVelocity());
  leftSensSim.SetRawPosition(leftPos);
  leftSensSim.SetVelocity(leftVel);
  rightSensSim.SetRawPosition(rightPos);
  rightSensSim.SetVelocity(rightVel);
  leftSim.SetRawRotorPosition(leftPos * kGearRatio);
  leftSim.SetRotorVelocity(leftVel * kGearRatio);
  rightSim.SetRawRotorPosition(rightPos * kGearRatio);
  rightSim.SetRotorVelocity(rightVel * kGearRatio);
  imuSim.SetRawYaw(m_driveSim.GetHeading().Degrees());

  /*
   * If a bumper is pressed, trigger the forward limit switch to test it,
   * if a trigger is pressed, trigger the reverse limit switch
   */
  leftSim.SetForwardLimit(joystick.GetLeftBumperButton());
  leftSim.SetReverseLimit(joystick.GetLeftTriggerAxis() > 0.5);
  rightSim.SetForwardLimit(joystick.GetRightBumperButton());
  rightSim.SetReverseLimit(joystick.GetRightTriggerAxis() > 0.5);
}

units::meter_t Robot::rotationsToMeters(units::turn_t rotations)
{
    /* Every radian of rotation, the wheel travels this many inches */
    constexpr auto wheelDistancePerRad = kWheelRadiusInches / 1_rad;
    /* Now apply gear ratio to input rotations */
    auto gearedRotations = rotations / kGearRatio;
    /* And multiply geared rotations by meters per rotation */
    return gearedRotations * wheelDistancePerRad;
}

units::turn_t Robot::metersToRotations(units::meter_t meters)
{
    /* Every radian of rotation, the wheel travels this many inches */
    constexpr auto wheelDistancePerRad = kWheelRadiusInches / 1_rad;
    /* Now get wheel rotations from input meters */
    auto wheelRadians = meters / wheelDistancePerRad;
    /* And multiply by gear ratio to get rotor rotations */
    return wheelRadians * kGearRatio;
}

units::meters_per_second_t Robot::rotationsToMetersVel(units::turns_per_second_t rotations)
{
    /* Every radian of rotation, the wheel travels this many inches */
    constexpr auto wheelDistancePerRad = kWheelRadiusInches / 1_rad;
    /* Now apply gear ratio to input rotations */
    auto gearedRotations = rotations / kGearRatio;
    /* And multiply geared rotations by meters per rotation */
    return gearedRotations * wheelDistancePerRad;
}

units::turns_per_second_t Robot::metersToRotationsVel(units::meters_per_second_t meters)
{
    /* Every radian of rotation, the wheel travels this many inches */
    constexpr auto wheelDistancePerRad = kWheelRadiusInches / 1_rad;
    /* Now get wheel rotations from input meters */
    auto wheelRadians = meters / wheelDistancePerRad;
    /* And multiply by gear ratio to get rotor rotations */
    return wheelRadians * kGearRatio;
}

#ifndef RUNNING_FRC_TESTS
int main()
{
  return frc::StartRobot<Robot>();
}
#endif
